### PR TITLE
Parameterized and multiple argument functionality enabled

### PR DESCRIPTION
**Pull request by samithkavishke**
## Purpose
$title

## Goals

- Parameterization of variables in the code base
- Supporting multiple arguments in a ballerina function
- Resolving bugs of supporting multiple function declaration

## Approach
New parameters were defined in the xml file to enable the multiple arguments. Singleton introduced for connector and the xml generation will be done at the LifeCycle phase. New parameters and methods introduced to connector to parameterize the variables.

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? no
 - Ran FindSecurityBugs plugin and verified report? no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/model/Connector.java [hunk 1]
```diff
@@ -60,10 +72,21 @@ public void setComponent(Component component) {
     }
 
     public String getType() {
-        return "connector";
+        return name;
+    }
+
+    public String getZipFileName() {
+        return this.name +"-"+TYPE_NAME+"-"+ this.version+ ".zip";
```
Comment (KavinduZoysa) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/model/Connector.java]: Please check the formatting.

Comment (KavinduZoysa) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/model/Connector.java]: Also other places.

File: compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/FunctionDefinitionAnalysisTask.java [hunk 2]
```diff
@@ -58,20 +57,17 @@ public void perform(SyntaxNodeAnalysisContext context) {
...
-        component.generateInstanceXml(connectorFolder);
-        component.generateTemplateXml(connectorFolder);
+//        Path connectorFolderPath = context.currentPackage().project().sourceRoot().resolve(Connector.TYPE_NAME);
+//        File connectorFolder = new File(connectorFolderPath.toUri());
+//        if (!connectorFolder.exists()) {
+//            connectorFolder.mkdir();
+//        }
+//        connector.setFolderPath(connectorFolderPath);
+//        component.generateInstanceXml(connectorFolder);
+//        component.generateTemplateXml(connectorFolder);
```
Comment (KavinduZoysa) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/FunctionDefinitionAnalysisTask.java]: Let's delete this.

File: compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/model/Connector.java (hunk 1 above)
Comment (samithkavishke) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/model/Connector.java]: Changed accordingly

File: compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/FunctionDefinitionAnalysisTask.java (hunk 2 above)
Comment (samithkavishke) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/FunctionDefinitionAnalysisTask.java]: removed

File: compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/BalCompilerLifeCycleTask.java [hunk 3]
```diff
@@ -1,22 +1,64 @@
...
 
 public class BalCompilerLifeCycleTask implements CompilerLifecycleTask<CompilerLifecycleEventContext> {
+
+    private static void createXmlFiles(Path connectorFolderPath, Connector connector) {
+        File connectorFolder = new File(connectorFolderPath.toUri());
+        if (!connectorFolder.exists()) {
+            connectorFolder.mkdir();
+        }
+
+        connector.generateInstanceXml(connectorFolder); //This need to be done after all the function definitions are analyzed
```
Comment (KavinduZoysa) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/BalCompilerLifeCycleTask.java]: Do we need this comment now?

File: compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/Utils.java [hunk 4]
```diff
@@ -4,32 +4,24 @@
 import com.github.jknack.handlebars.Template;
 import io.ballerina.stdlib.mi.plugin.model.ModelElement;
 
-import java.io.BufferedReader;
-import java.io.FileWriter;
-import java.io.IOException;
-import java.io.InputStream;
-import java.io.InputStreamReader;
+import java.io.*;
```
Comment (KavinduZoysa) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/Utils.java]: Individual imports are needed.

Comment (KavinduZoysa) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/Utils.java]: Please check other places too.

File: compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/BalCompilerLifeCycleTask.java [hunk 5]
```diff
@@ -1,22 +1,64 @@
...
         Path sourcePath = Paths.get(sourcePathStr);
-        Path destinationPath = compilerLifecycleEventContext.currentPackage().project().sourceRoot().resolve(Constants.CONNECTOR);
 
+        PackageDescriptor descriptor = context.currentPackage().manifest().descriptor();
+        String moduleName = descriptor.name().value();
+
+        Path destinationPath = context.currentPackage().project().sourceRoot().resolve(Connector.TYPE_NAME);
+
+        Connector connector = Connector.getConnector();
+        connector.setName(moduleName);
```
Comment (KavinduZoysa) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/BalCompilerLifeCycleTask.java]: Let's define `moduleName` right above this line or use it without defining an extra variable.

File: compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/BalCompilerLifeCycleTask.java [hunk 6]
```diff
@@ -1,22 +1,64 @@
...
+        // Generate all the xml files
+        createXmlFiles(destinationPath, connector);
+
+        // Copy the resources from the resources folder to the connector folder
+        try {
+            URI jarPath = getClass().getProtectionDomain().getCodeSource().getLocation().toURI();
+            Utils.copyResources(getClass().getClassLoader(), destinationPath, jarPath);
+        } catch (IOException | URISyntaxException e) {
+            throw new RuntimeException(e);
+        }
```
Comment (KavinduZoysa) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/BalCompilerLifeCycleTask.java]: We can combine this try catch with the below one.

File: compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/model/Component.java [hunk 7]
```diff
@@ -7,15 +7,17 @@
 
 public class Component extends ModelElement {
 
+    public static final String TYPE_NAME = "component";
     private String name = "dummy";
```
Comment (KavinduZoysa) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/model/Component.java]: Let's change this value.

File: compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/model/Connector.java [hunk 8]
```diff
@@ -7,18 +7,25 @@
...
-    private String name = "connector";
-
+    public static final String TYPE_NAME = "connector";
+    private static final Connector connector = new Connector();
+    private final ArrayList<Component> components = new ArrayList<>();
+    private String name = "";
     private String description = "helps to connect with external systems";
-
     private String iconPath = "icon/icon-small.gif";
+    private String packageName = "";
```
Comment (KavinduZoysa) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/model/Connector.java]: Let's remove unused variables. Please check all other classes.

File: compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/model/Component.java (hunk 7 above)
Comment (samithkavishke) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/model/Component.java]: change variable name or?

Comment (KavinduZoysa) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/model/Component.java]: value

Comment (samithkavishke) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/model/Component.java]: removed that initialization

File: compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/BalCompilerLifeCycleTask.java (hunk 5 above)
Comment (samithkavishke) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/BalCompilerLifeCycleTask.java]: Temporary variable removed

File: compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/BalCompilerLifeCycleTask.java [hunk 9]
```diff
@@ -1,35 +1,60 @@
...
-        }
+        PackageDescriptor descriptor = context.currentPackage().manifest().descriptor();
+        Path destinationPath = context.currentPackage().project().sourceRoot().resolve(Connector.TYPE_NAME);
+
+        Connector connector = Connector.getConnector();
+        connector.setName(descriptor.name().value());
+        connector.setVersion(descriptor.version().value().toString());
+
+        // Generate all the xml files
+        createXmlFiles(destinationPath, connector);
```
Comment (KavinduZoysa) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/BalCompilerLifeCycleTask.java]: Let's remove the comment and rename `generateXmlFiles`.

File: compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/BalCompilerLifeCycleTask.java [hunk 10]
```diff
@@ -1,35 +1,60 @@
...
         try {
-            // TODO : Parameterize the connector zip name
-            Utils.zipFolder(destinationPath, "BallerinaTransformer-connector-1.0-SNAPSHOT.zip");
-            Utils.deleteDirectory(destinationPath);
-        } catch (IOException e) {
+            URI jarPath = getClass().getProtectionDomain().getCodeSource().getLocation().toURI();
+            Utils.copyResources(getClass().getClassLoader(), destinationPath, jarPath); // Copy the resources from the resources folder to the connector folder
+            Files.copy(sourcePath, destinationPath.resolve("lib").resolve(sourcePath.getFileName())); // Copy the Ballerina JAR file to the connector folder
+            Utils.zipFolder(destinationPath, connector.getZipFileName()); // Zip the connector folder
+            Utils.deleteDirectory(destinationPath); // delete the folder
```
Comment (KavinduZoysa) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/BalCompilerLifeCycleTask.java]: Let's delete these comments.

File: compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/Utils.java [hunk 11]
```diff
@@ -4,32 +4,24 @@
...
 import java.nio.charset.StandardCharsets;
 import java.nio.file.FileSystem;
-import java.nio.file.FileSystems;
-import java.nio.file.Files;
-import java.nio.file.Path;
-import java.nio.file.Paths;
-import java.nio.file.SimpleFileVisitor;
-import java.nio.file.FileVisitResult;
-
+import java.nio.file.*;
```
Comment (KavinduZoysa) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/mi/plugin/Utils.java]: We need to have individual imports.

**Review by KavinduZoysa — APPROVED** [review]

**Pull request merged**